extensions/smallrye-graphql/runtime/src/main/java/io/quarkus/smallrye/graphql/runtime/SmallRyeGraphQLSubscriptionHandler.java

```diff
@@ -0,0 +1,180 @@
+        @Override
+        public void handle(Void e) {
+            unsubscribe();
+        }
+    }
+
+    private class ExceptionHandler implements Handler<Throwable> {
+        @Override
+        public void handle(Throwable e) {
+            log.error(e.getMessage());
+            unsubscribe();
+        }
+    }
+
+    public void unsubscribe() {
+        Subscriptions.cancel(subscriptionRef);
+        subscriptionRef.set(null);
+    }
+
+    private class TextMessageHandler implements Handler<String> {
+        private final SmallRyeGraphQLSubscriptionSubscriber smallRyeGraphQLSubscriptionSubscriber;
+
+        TextMessageHandler(final ServerWebSocket serverWebSocket) {
+            this.smallRyeGraphQLSubscriptionSubscriber = new SmallRyeGraphQLSubscriptionSubscriber(serverWebSocket);
+        }
+
+        @Override
+        public void handle(String message) {
+            JsonObject jsonInput = inputToJsonObject(message);
+            ExecutionResponse executionResponse = getExecutionService()
+                    .execute(jsonInput);
+
+            ExecutionResult executionResult = executionResponse.getExecutionResult();
+            if (executionResult != null) {
+                // If there is error on the query, we can not start a subscription
+                if (executionResult.getErrors() != null && !executionResult.getErrors().isEmpty()) {
+                    smallRyeGraphQLSubscriptionSubscriber.onNext(executionResult);
+                    smallRyeGraphQLSubscriptionSubscriber.closeWebSocket();
+                } else {
+                    Publisher<ExecutionResult> stream = executionResponse.getExecutionResult()
+                            .getData();
+
+                    if (stream != null) {
+
+                        Multi<ExecutionResult> multiStream = Multi.createFrom().publisher(stream);
+
+                        multiStream.onFailure().recoverWithItem(failure -> {
+                            // TODO: Below must move to SmallRye, and once 1.2.1 is releaes we can remove it here
+                            //       Once in SmallRye, this will also follow the propper error rule (show/hide) and add more details.
+                            return new ExecutionResultImpl(GraphqlErrorBuilder.newError()
+                                    .message(failure.getMessage())
+                                    .build());
+                        }).subscribe(smallRyeGraphQLSubscriptionSubscriber);
+                    }
+                }
+            }
+        }
+    }
+
+    private class SmallRyeGraphQLSubscriptionSubscriber implements Subscriber<ExecutionResult> {
+        private final ServerWebSocket serverWebSocket;
+
+        public SmallRyeGraphQLSubscriptionSubscriber(ServerWebSocket serverWebSocket) {
+            this.serverWebSocket = serverWebSocket;
+        }
+
+        @Override
+        public void onSubscribe(Subscription s) {
+            subscriptionRef.set(s);
```

Comment:
if (subscriptionRef.compareAndSet(null, s) { s.request(1); }
else {
    s.cancel(); // You can't handle multiple subscribers
}

Reply:
Done